zebra/zapi_msg.c

```diff
@@ -748,7 +748,8 @@ static int route_notify_internal(const struct prefix *p, int type,
 			   vrf_id);
 	}
 
-	s = stream_new(ZEBRA_MAX_PACKET_SIZ);
+	s = stream_new(200);
```

Comment:
Don't we have a definition of how small this is? 200 looks a bit arbitrary, maybe a comment explaining this value?

Reply:
yeah, that's a good point - I saw that we were allocating 16K here, but just encoding a prefix and a couple of ints. what I was seeing was that the average 'stream' size was affected by this, as zebra queued these notifications.

Reply:
I would prefer we allocate something that makes sense.  I've left this at 16k because when I did back of the math calculations of nexthop sizes with labels and all the other data that can be encoded we needed something fairly large (see 256 x ecmp w/ labels on each ecmp as an example ).  200 bytes just ain't gonna cut it.

I've even started down the path of trying to figure out ways to intelligently allocate it.  I would prefer more thought here.

Reply:
but this doesn't include nexthops - it isn't redistribution, or nht - it's just the prefix, and the status? nexthops aren't even available in the api to this function. I agree that it wouldn't have been good to have tried a small buffer for paths that might include nexthops.

Reply:
I did not convey what I was trying to say very well.  There are a large number of data packets being sent back and forth between the daemons and zebra.  Perhaps we need a generic way to convey what the stream size should be on a per zapi type basis?  Instead of doing this one off.  What about redistribution events when you ahve a full table?

Reply:
Agree it'd be nice to have a clean way to estimate the needed buffer size. For the zapi messages that use an intermediate struct like zapi_route or zapi_labels, that's pretty natural. but many messages just open-code their marshall/unmarshall requirements, often in multiple daemons. In those cases, it's going to be difficult even to publicize how the size should be maintained, and it's going to be impossible to enforce it. Should we convert all the open-coded messages to use intermediate structs?

The paths for zebra and for other daemons also differ: zebra enqueues 'stream' objects for sending in an io pthread, so each message needs to allocate its own object. Everyone else uses a single stream built into the zclient struct, and sends each message immediately. That zclient stream sort of tries to max-size itself at init time.

I was only trying here to avoid enqueuing MAX_PACKET-sized objects when that was ... a couple of orders of magnitude larger than we needed. I don't think that's really a big part of the memory-footprint issue, at least not in this artificial test/CI situation - just seemed like something worth attending to.